### PR TITLE
GUACAMOLE-300: Support posixGroup in LDAP Authentication and Group-based Session Admission

### PR DESCRIPTION
**Pull request by mlewissmith**
To support groups where group members are specified by usercode, rather than by DN
- new guacamole.property 'ldap-member-attribute-type'.  Can take value "dn" (as now) or "uid"

### CODE REVIEW

File: extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/group/UserGroupService.java [hunk 1]
```diff
@@ -165,14 +165,44 @@ private String getGroupSearchFilter() throws GuacamoleException {
         if (groupBaseDN == null)
             return Collections.emptyList();
 
+        // memberAttribute specified in properties could contain DN or username 
+        String memberAttributeType = confService.getMemberAttributeType();
+        String userID = userDN;
+        switch (memberAttributeType) {
```
Comment (necouchman) [inline extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/group/UserGroupService.java]: Might be worth doing a new `enum` type for this and a `GuacamoleProperty` that validates it.

**Comment by mike-jumper** [comment]
> ... I suggest doing an `enum` and `GuacamoleProperty` to cover this new property so that it can be validated correctly.

+1

I also suggest holding on merge until after #345, as this will likely need to be rebased and slightly modified following the migration from JLDAP to the Apache Directory API. These changes will otherwise end up needing to be rewritten during merge conflict resolution, which will be awkward.

**Comment by mlewissmith** [comment]
yes. of course it should be an enum...

File: extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/group/UserGroupService.java (hunk 1 above)
Comment (necouchman) [reply extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/group/UserGroupService.java]: Looks much better with the `enum`, here, but, now I think we can probably get rid of the `switch()` statement, no?  Based on the way the logic is constructed, here, I think you can just do the `String userID = userDN;` assignment, and then do a `if (memberAttributeType == MemberAttributeType.UID)` with and re-assign as necessary?

**Comment by mlewissmith** [comment]
switch is gone as requested.  I'm happy to wait and deal with the API changes when the time comes

**Comment by mlewissmith** [comment]
Sorry -- forgot to tag the commit messages appropriately.  I hope force-push doesn't cause problems

File: extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/ConfigurationService.java [hunk 2]
```diff
@@ -396,4 +396,21 @@ public String getMemberAttribute() throws GuacamoleException {
         );
     }
 
+    /**
+     * Returns a string indicating the type of data contained in the LDAP
```
Comment (necouchman) [inline extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/ConfigurationService.java]: No longer returns a `String` - is now a `MemberAttributeType`

File: extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/ConfigurationService.java [hunk 3]
```diff
@@ -396,4 +396,21 @@ public String getMemberAttribute() throws GuacamoleException {
...
+     * attribute used to enumerate members in a group, "dn" by default.
+     *
+     * @return
+     *     The type of data contained in the LDAP attribute used to enumerate
+     *     members in a group, as configured in guacamole.properties
+     *
+     * @throws GuacamoleException
+     *     If guacamole.properties connect be parsed.
```
Comment (necouchman) [inline extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/ConfigurationService.java]: connect -> cannot

File: extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/MemberAttributeTypeProperty.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        // dn
+        if (value.equals("dn"))
+            return MemberAttributeType.DN;
+
+        // uid
+        if (value.equals("uid"))
+            return MemberAttributeType.UID;
+
+        // The provided value is not legal
+        throw new GuacamoleServerException("Encryption method must be one of \"dn\" or \"uid\".");
```
Comment (necouchman) [inline extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/MemberAttributeTypeProperty.java]: Copypasta, I think - this is not the encryption method ;-).

File: extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/ConfigurationService.java (hunk 3 above)
Comment (mlewissmith) [reply extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/ConfigurationService.java]: ... also copypasta, from getMemberAttribute()

**Comment by necouchman** [comment]
@mlewissmith: Okay, changes for the LDAP API in use (JLDAP -> Apache) have been committed, so should be safe to clean this up, now, and take another look.

**Comment by mlewissmith** [comment]
I may have messed up the rebase.  Do you want me to close this PR and start again?

**Comment by necouchman** [comment]
> I may have messed up the rebase. Do you want me to close this PR and start again?

Up to you - you might be able to rebase, again, on top of the current master, and pull out all of the changes except yours, or you can just close it and open a new one.  It does seem to be a little messed up - 6K lines of code changed :-).

**Pull request closed without merging**